### PR TITLE
fix(26141): String type autocomplete doesn't work with paren

### LINKED ISSUE #26141: String type autocomplete doesn't work with paren
**TypeScript Version:**  3.1.0-dev.201xxxxx

**Search Terms:**
completion, paren
**Code**

```ts
type A = 'AAAAA' | 'BBBBBB';
type B<T extends A> = {};
type C = B<('/**/')>
```

**Expected behavior:**
`'AAAAA'` or `'BBBBBB'`

**Actual behavior:**
no completions

**Related Issues:** 
#25839 

completion cannot walkup useless node (paren type)

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #26141

### CODE REVIEW

File: src/services/stringCompletions.ts [hunk 1]
```diff
@@ -114,6 +114,10 @@ namespace ts.Completions.StringCompletions {
                 switch (parent.parent.kind) {
                     case SyntaxKind.TypeReference:
                         return { kind: StringLiteralCompletionKind.Types, types: getStringLiteralTypes(typeChecker.getTypeArgumentConstraint(parent as LiteralTypeNode)), isNewIdentifier: false };
+                    case SyntaxKind.ParenthesizedType: {
+                        const typeNode = findAncestor(parent.parent, n => !isParenthesizedTypeNode(n.parent)) as ParenthesizedTypeNode;
+                        return { kind: StringLiteralCompletionKind.Types, types: getStringLiteralTypes(typeChecker.getTypeArgumentConstraint(typeNode)), isNewIdentifier: false };
```
Comment (DanielRosenwasser) [inline src/services/stringCompletions.ts:119]: This doesn't seem right - you don't necessarily know whether you're in a "type argument constraint" position. That's the issue. For example, you could have

```ts
type Foo = Bar[  (  (  "/**/"  )  )  ];
```

in which case, you actually want to give results from the `IndexedAccessType` case, right?

Comment (a-tarasyuk) [reply src/services/stringCompletions.ts:119]: The fix handles the case `<(( "/**/" ))>` which was described in the bug description, it doesn't affect completion for `IndexedAccessType`.

File: src/services/stringCompletions.ts [hunk 2]
```diff
@@ -114,6 +114,14 @@ namespace ts.Completions.StringCompletions {
                 switch (parent.parent.kind) {
                     case SyntaxKind.TypeReference:
                         return { kind: StringLiteralCompletionKind.Types, types: getStringLiteralTypes(typeChecker.getTypeArgumentConstraint(parent as LiteralTypeNode)), isNewIdentifier: false };
+                    case SyntaxKind.ParenthesizedType: {
+                        const typeNode = findAncestor(parent.parent, n => !isParenthesizedTypeNode(n.parent)) as ParenthesizedTypeNode;
+                        const typeArguments = (typeNode.parent as NodeWithTypeArguments).typeArguments;
```
Comment (DanielRosenwasser) [inline src/services/stringCompletions.ts:119]: What about

```ts
Foo[("hello")]
```

?

Comment (DanielRosenwasser) [reply src/services/stringCompletions.ts:119]: I think in short, you should really consider wrapping this whole thing in a loop where you check the parents each time.

File: src/services/stringCompletions.ts [hunk 3]
```diff
@@ -114,6 +114,14 @@ namespace ts.Completions.StringCompletions {
...
                         return { kind: StringLiteralCompletionKind.Types, types: getStringLiteralTypes(typeChecker.getTypeArgumentConstraint(parent as LiteralTypeNode)), isNewIdentifier: false };
+                    case SyntaxKind.ParenthesizedType: {
+                        const typeNode = findAncestor(parent.parent, n => !isParenthesizedTypeNode(n.parent)) as ParenthesizedTypeNode;
+                        const typeArguments = (typeNode.parent as NodeWithTypeArguments).typeArguments;
+                        if (typeArguments && find(typeArguments, typeArgument => typeArgument === typeNode)) {
+                            return { kind: StringLiteralCompletionKind.Types, types: getStringLiteralTypes(typeChecker.getTypeArgumentConstraint(typeNode)), isNewIdentifier: false };
+                        }
+                        return undefined;
+                    }
                     case SyntaxKind.IndexedAccessType:
```
Comment (DanielRosenwasser) [inline src/services/stringCompletions.ts:125]: #40322

File: src/services/stringCompletions.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/stringCompletions.ts:119]: @DanielRosenwasser Not sure that I got your idea :(. I thought about the issue like so, `type T3 = T2<(((((/**/)))))>;` parent is `ParenthesizedType`, we need to skip all `ParenthesizedType` to get correct type argument. Or am I wrong?

`Foo[("hello")]` it is not `ParenthesizedType`. this case will not be handled.

**Review by sandersn — CHANGES_REQUESTED** [review]
Would it work to use skipParentheses instead of recursion?
Also I don't understand why ImportType and UnionType need to return undefined for non-string-literals.

**Pull request merged**